File: sdk/python/costorah/instrumentation/cohere.py
"""
CohereInstrumentor — automatic usage capture for the official `cohere`
Python package's `ClientV2.chat`/`chat_stream` (and the async
`AsyncClientV2` equivalents).

    import cohere
    from costorah.instrumentation import CohereInstrumentor

    CohereInstrumentor().instrument()

    client = cohere.ClientV2(api_key="...")
    client.chat(model="command-r-plus", messages=[...])
"""

from __future__ import annotations

import functools
import time
from typing import Any

from costorah._logging import get_logger
from costorah._util import generate_request_id
from costorah.instrumentation._streaming import InstrumentedAsyncStream, InstrumentedSyncStream
from costorah.instrumentation._submission import submit
from costorah.instrumentation.base import BaseInstrumentor, ExtractedUsage, InstrumentationError
from costorah.instrumentation.pricing import calculate_cost
from costorah.types import UsageStatus

_log = get_logger(__name__)


def _extract(response: Any) -> dict[str, Any]:
    usage = getattr(response, "usage", None)
    tokens = getattr(usage, "tokens", None) if usage else None
    if tokens is None:
        return {}
    return {
        "input_tokens": getattr(tokens, "input_tokens", 0) or 0,
        "output_tokens": getattr(tokens, "output_tokens", 0) or 0,
    }


def _aggregate_stream(events: list[Any]) -> dict[str, Any]:
    # The terminal "message-end" event carries the final usage/response.
    for event in reversed(events):
        response = getattr(event, "response", None)
        raw = _extract(response) if response is not None else _extract(event)
        if raw:
            return raw
    return {}


class CohereInstrumentor(BaseInstrumentor):
    name = "cohere"

    def extract_usage(self, response: Any) -> dict[str, Any]:
        return _extract(response)

    def normalize(
        self,
        raw_usage: dict[str, Any],
        *,
        model: str,
        latency_ms: int,
        status: UsageStatus,
        request_id: str | None = None,
    ) -> ExtractedUsage:
        input_tokens = int(raw_usage.get("input_tokens", 0) or 0)
        output_tokens = int(raw_usage.get("output_tokens", 0) or 0)
        cost, estimated = (
            calculate_cost("cohere", model, input_tokens, output_tokens)
            if self.calculate_cost_enabled
            else (0.0, False)
        )
        metadata: dict[str, Any] = {}
        if self.capture_metadata:
            metadata["cost_estimated"] = estimated
        return ExtractedUsage(
            provider="cohere",
            model=model,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            cost=cost,
            latency_ms=latency_ms,
            status=status,
            request_id=request_id or generate_request_id(),
            metadata=metadata,
        )

    def _apply_patches(self) -> None:
        try:
            from cohere import AsyncClientV2, ClientV2
        except ImportError as exc:
            raise InstrumentationError(
                "The 'cohere' package is not installed. Install it with "
                "`pip install cohere` to use this instrumentor."
            ) from exc

        # `chat`/`chat_stream` are defined on a base class (V2Client), not
        # directly on ClientV2/AsyncClientV2 — use getattr() (MRO-aware)
        # rather than __dict__ to find the real original, matching what
        # BaseInstrumentor._patch() already does internally when recording
        # the patch to restore.
        self._patch(ClientV2, "chat", self._make_wrapper(ClientV2.chat, "sync"))
        self._patch(AsyncClientV2, "chat", self._make_wrapper(AsyncClientV2.chat, "async"))
        self._patch(
            ClientV2,
            "chat_stream",
            self._make_wrapper(ClientV2.chat_stream, "sync_stream"),
        )
        self._patch(
            AsyncClientV2,
            "chat_stream",
            self._make_wrapper(AsyncClientV2.chat_stream, "async_stream"),
        )

    def _make_wrapper(self, original: Any, mode: str) -> Any:
        instrumentor = self

        def submit_result(model: str, raw: dict[str, Any], elapsed_ms: int, status: str) -> None:
            usage = instrumentor.normalize(
                raw,
                model=model,
                latency_ms=elapsed_ms,
                status=status,  # type: ignore[arg-type]
            )
            instrumentor._record_captured()
            submit(usage, client=instrumentor._client)

        if mode == "async":

            @functools.wraps(original)
            async def async_wrapper(self_client: Any, *args: Any, **kwargs: Any) -> Any:
                model = kwargs.get("model", "unknown")
                start = time.perf_counter()
                try:
                    result = await original(self_client, *args, **kwargs)
                except Exception:
                    submit_result(model, {}, instrumentor._elapsed_ms(start), "error")
                    raise
                submit_result(model, _extract(result), instrumentor._elapsed_ms(start), "success")
                return result

            return async_wrapper

        if mode == "sync_stream":

            @functools.wraps(original)
            def sync_stream_wrapper(self_client: Any, *args: Any, **kwargs: Any) -> Any:
                model = kwargs.get("model", "unknown")
                start = time.perf_counter()
                result = original(self_client, *args, **kwargs)

                def on_complete(
                    chunks: list[Any], elapsed_ms: int, error: Exception | None
                ) -> None:
                    raw = _aggregate_stream(chunks)
                    submit_result(model, raw, elapsed_ms, "error" if error else "success")

                return InstrumentedSyncStream(iter(result), start, on_complete)

            return sync_stream_wrapper

        if mode == "async_stream":

            @functools.wraps(original)
            async def async_stream_wrapper(self_client: Any, *args: Any, **kwargs: Any) -> Any:
                model = kwargs.get("model", "unknown")
                start = time.perf_counter()
                result = await original(self_client, *args, **kwargs)

                def on_complete(
                    chunks: list[Any], elapsed_ms: int, error: Exception | None
                ) -> None:
                    raw = _aggregate_stream(chunks)
                    submit_result(model, raw, elapsed_ms, "error" if error else "success")

                return InstrumentedAsyncStream(result.__aiter__(), start, on_complete)

            return async_stream_wrapper

        @functools.wraps(original)
        def sync_wrapper(self_client: Any, *args: Any, **kwargs: Any) -> Any:
            model = kwargs.get("model", "unknown")
            start = time.perf_counter()
            try:
                result = original(self_client, *args, **kwargs)
            except Exception:
                submit_result(model, {}, instrumentor._elapsed_ms(start), "error")
                raise
            submit_result(model, _extract(result), instrumentor._elapsed_ms(start), "success")
            return result

        return sync_wrapper

File: sdk/javascript/src/instrumentation/cohere.ts
/**
 * CohereInstrumentor — automatic usage capture for the official
 * `cohere-ai` npm package's `chat`/`chatStream`.
 *
 *     import { CohereClientV2 } from "cohere-ai";
 *     import { CohereInstrumentor } from "@costorah/sdk";
 *
 *     const client = new CohereClientV2({ token: "..." });
 *     new CohereInstrumentor().instrument(client);
 *
 *     await client.chat({ model: "command-r-plus", messages: [] });
 *
 * Like `@google/genai`, `cohere-ai`'s `CohereClientV2.chat`/`chatStream`
 * are own-instance bound-function properties (confirmed via
 * `Object.getOwnPropertyDescriptor(client, "chat")` showing
 * `{ value: [Function: bound chat], writable: true, configurable: true }`
 * directly on the instance), not prototype methods — there is no shared
 * prototype to patch once process-wide. This instrumentor requires the
 * specific client instance to wrap, via `instrument(client)` instead of a
 * zero-arg `instrument()`.
 */

import type { UsageStatus } from "../types.js";
import { BaseInstrumentor, InstrumentationError, makeExtractedUsage } from "./base.js";
import type { ExtractedUsage } from "./base.js";
import { calculateCost } from "./pricing.js";
import { instrumentedAsyncStream } from "./streaming.js";
import { submit } from "./submission.js";

interface CohereUsage {
  tokens?: { inputTokens?: number; outputTokens?: number };
}

function extract(response: unknown): Record<string, unknown> {
  const usage = (response as { usage?: CohereUsage } | undefined)?.usage;
  const tokens = usage?.tokens;
  if (!tokens) return {};
  return {
    inputTokens: tokens.inputTokens ?? 0,
    outputTokens: tokens.outputTokens ?? 0,
  };
}

/** The terminal "message-end" stream event carries the final response
 * (and usage) — walk backwards from the end to find it. */
function aggregateStream(events: unknown[]): Record<string, unknown> {
  for (let i = events.length - 1; i >= 0; i--) {
    const event = events[i];
    const response = (event as { response?: unknown })?.response;
    const raw = response !== undefined ? extract(response) : extract(event);
    if (Object.keys(raw).length > 0) return raw;
  }
  return {};
}

function generateRequestId(): string {
  const random =
    typeof crypto !== "undefined" && "randomUUID" in crypto
      ? crypto.randomUUID().replace(/-/g, "")
      : `${Date.now().toString(16)}${Math.random().toString(16).slice(2)}`;
  return `sdk_js_instr_${random}`;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ChatMethod = (...args: any[]) => unknown;

interface CohereClient {
  chat: ChatMethod;
  chatStream: ChatMethod;
}

export class CohereInstrumentor extends BaseInstrumentor {
  readonly name = "cohere";

  private target: CohereClient | undefined;
  private originalChat: ChatMethod | undefined;
  private originalChatStream: ChatMethod | undefined;

  extractUsage(response: unknown): Record<string, unknown> {
    return extract(response);
  }

  normalize(
    rawUsage: Record<string, unknown>,
    context: { model: string; latencyMs: number; status: UsageStatus; requestId?: string },
  ): ExtractedUsage {
    const inputTokens = Number(rawUsage.inputTokens ?? 0);
    const outputTokens = Number(rawUsage.outputTokens ?? 0);
    const { cost, estimated } = this.calculateCostEnabled
      ? calculateCost("cohere", context.model, inputTokens, outputTokens)
      : { cost: 0, estimated: false };
    const metadata: Record<string, unknown> = {};
    if (this.captureMetadata) metadata.costEstimated = estimated;

    return makeExtractedUsage({
      provider: "cohere",
      model: context.model,
      inputTokens,
      outputTokens,
      cost,
      latencyMs: context.latencyMs,
      status: context.status,
      requestId: context.requestId ?? generateRequestId(),
      metadata,
    });
  }

  /** CohereInstrumentor cannot patch a shared prototype (see module
   * docstring) — pass the specific `CohereClientV2` instance to wrap. */
  instrument(client?: CohereClient): void {
    if (!this.enabled) return;
    if (this.isInstrumented()) return;
    if (!client || typeof client.chat !== "function") {
      throw new InstrumentationError(
        "CohereInstrumentor.instrument(client) requires a CohereClientV2 instance — " +
          "call `new CohereInstrumentor().instrument(client)`.",
      );
    }
    this.target = client;
    this.originalChat = client.chat.bind(client);
    this.originalChatStream = client.chatStream.bind(client);
    client.chat = this.wrapChat(this.originalChat);
    client.chatStream = this.wrapChatStream(this.originalChatStream);
    this.markInstrumented();
  }

  uninstrument(): void {
    if (!this.isInstrumented()) return;
    if (this.target && this.originalChat && this.originalChatStream) {
      this.target.chat = this.originalChat;
      this.target.chatStream = this.originalChatStream;
    }
    this.target = undefined;
    this.originalChat = undefined;
    this.originalChatStream = undefined;
    this.markUninstrumented();
  }

  protected applyPatches(): void {
    throw new InstrumentationError(
      "CohereInstrumentor requires a client instance: call instrument(client), not instrument().",
    );
  }

  protected removePatches(): void {
    // Unused: uninstrument() is fully overridden above.
  }

  private async submitResult(
    model: string,
    raw: Record<string, unknown>,
    latencyMs: number,
    status: UsageStatus,
  ): Promise<void> {
    const usage = this.normalize(raw, { model, latencyMs, status });
    this.recordCaptured();
    await submit(usage);
  }

  private wrapChat(original: ChatMethod): ChatMethod {
    return async (...args: unknown[]) => {
      const params = (args[0] ?? {}) as { model?: string };
      const model = params.model ?? "unknown";
      const start = Date.now();
      let result: unknown;
      try {
        result = await original(...args);
      } catch (err) {
        await this.submitResult(model, {}, Date.now() - start, "error");
        throw err;
      }
      await this.submitResult(model, extract(result), Date.now() - start, "success");
      return result;
    };
  }

  private wrapChatStream(original: ChatMethod): ChatMethod {
    return async (...args: unknown[]) => {
      const params = (args[0] ?? {}) as { model?: string };
      const model = params.model ?? "unknown";
      const start = Date.now();
      const result = await original(...args);
      return instrumentedAsyncStream(
        result as AsyncIterable<unknown>,
        start,
        async (chunks, elapsedMs, error) => {
          const raw = aggregateStream(chunks);
          await this.submitResult(model, raw, elapsedMs, error ? "error" : "success");
        },
      );
    };
  }
}
